pinot-common/src/main/java/org/apache/pinot/common/function/sql/PinotFunctionRegistry.java

```diff
@@ -0,0 +1,237 @@
+    //   1. register those that are supported in multistage in addition to calcite standard opt table.
+    //   2. register special handling that differs from calcite standard.
+    for (TransformFunctionType transformFunctionType : TransformFunctionType.values()) {
+      if (transformFunctionType.getSqlKind() != null) {
+        // 1. Register the transform function with Calcite
+        registerTransformFunction(transformFunctionType.getName(), transformFunctionType);
+        // 2. Register the transform function with Calcite on all alternative names
+        List<String> alternativeFunctionNames = transformFunctionType.getAlternativeNames();
+        for (String alternativeFunctionName : alternativeFunctionNames) {
+          registerTransformFunction(alternativeFunctionName, transformFunctionType);
+        }
+      }
+    }
+  }
+
+  public static void init() {
+  }
+
+  @VisibleForTesting
+  public static void registerFunction(Method method, boolean nullableParameters) {
+    registerFunction(method, Collections.singleton(method.getName()), nullableParameters);
+  }
+
+  public static NameMultimap<PinotFunction> getFunctionMap() {
+    return FUNCTION_MAP;
+  }
+
+  public static NameMultimap<SqlOperator> getOperatorMap() {
+    return OPERATOR_MAP;
+  }
+
+  @Nullable
+  public static PinotScalarFunction getScalarFunction(SqlOperatorTable operatorTable, RelDataTypeFactory typeFactory,
+      String functionName, List<DataSchema.ColumnDataType> argTypes) {
+    List<RelDataType> relArgTypes = convertArgumentTypes(typeFactory, argTypes);
+    SqlOperator sqlOperator = SqlUtil.lookupRoutine(operatorTable, typeFactory,
+        new SqlIdentifier(functionName, SqlParserPos.QUOTED_ZERO), relArgTypes, null, null, SqlSyntax.FUNCTION,
+        SqlKind.OTHER_FUNCTION, SqlNameMatchers.withCaseSensitive(false), true);
+    if (sqlOperator instanceof SqlUserDefinedFunction) {
+      Function function = ((SqlUserDefinedFunction) sqlOperator).getFunction();
+      if (function instanceof PinotScalarFunction) {
+        return (PinotScalarFunction) function;
+      }
+    }
+    return null;
+  }
+
+  private static void registerFunction(Method method, Set<String> alias, boolean nullableParameters) {
+    if (method.getAnnotation(Deprecated.class) == null) {
+      for (String name : alias) {
+        registerCalciteNamedFunctionMap(name, method, nullableParameters);
+      }
+    }
+  }
+
+  private static void registerCalciteNamedFunctionMap(String name, Method method, boolean nullableParameters) {
+    FUNCTION_MAP.put(name, new PinotScalarFunction(name, method, nullableParameters));
+  }
+
+  private static List<RelDataType> convertArgumentTypes(RelDataTypeFactory typeFactory,
+      List<DataSchema.ColumnDataType> argTypes) {
+    return argTypes.stream().map(type -> toRelType(typeFactory, type)).collect(Collectors.toList());
+  }
+
+  private static RelDataType toRelType(RelDataTypeFactory typeFactory, DataSchema.ColumnDataType dataType) {
```

Comment:
Note to reviewer: 
this and several other type conversion utils should be unified, we will follow up in separate PRs